### PR TITLE
Add support for HSM-based signing

### LINKED ISSUE #18: Using an HSM instead of file for CUSTOMER_KEY_FILE_PEM?
Hello, this new rpi-sb-provisioner looks super cool!

I wondered if there is any way planned to use an HSM for the signing key instead of just having it in a file ?

### PR DESCRIPTION
**Pull request by tdewey-rpi**
Add rudimentary support for using a PKCS11 engine to provide the private key for the signed boot component.

This does not impact the device-unique key that is generated as part of provisioning.

Where the PKCS11_NAME and FILE_PEM are both supplied, favour the PKCS11 name.

Fixes #18

### CODE REVIEW

File: host-support/terminal-functions.sh [hunk 1]
```diff
@@ -130,4 +130,21 @@ get_fastboot_config_file() {
...
         echo "/var/lib/rpi-sb-provisioner/boot_ramdisk_config.txt"
     fi
+}
+
+get_signing_directives() {
+    if [ -n "${CUSTOMER_KEY_PKCS11_NAME}" ]; then
+        echo "${CUSTOMER_KEY_PKCS11_NAME} -engine pkcs11 -keyform engine"
+    else
+        if [ -n "${CUSTOMER_KEY_FILE_PEM}" ]; then
+            if [ -f "${CUSTOMER_KEY_FILE_PEM}" ]; then
```
[the remark this answers is not in the document]
Comment (tdewey-rpi) [reply host-support/terminal-functions.sh]: Fixed in next commit

File: device-triage/triage.sh [hunk 2]
```diff
@@ -21,8 +21,8 @@ if [ -z "${RPI_DEVICE_BOOTLOADER_CONFIG_FILE}" ]; then
     RPI_DEVICE_BOOTLOADER_CONFIG_FILE=/var/lib/rpi-sb-provisioner/bootloader.default
 fi
 
-if [ -z "${CUSTOMER_KEY_FILE_PEM}" ]; then
-    echo "You must provide a customer key file in the environment via CUSTOMER_KEY_FILE_PEM"
+if [ -z "${CUSTOMER_KEY_FILE_PEM}" ] || [ -z "${CUSTOMER_KEY_PKCS11_NAME}" ]; then
```
Comment (roliver-rpi) [inline device-triage/triage.sh:24]: I believe this should be `&&`, i.e. if neither of them are set we've got an error condition.

**Comment by roliver-rpi** [comment]
Do we have any additional dependencies, e.g. `libengine-pkcs11-openssl`?

**Comment by tdewey-rpi** [comment]
> Do we have any additional dependencies, e.g. `libengine-pkcs11-openssl`?

Unknown at this time - but that would make sense. I need to set up a PKCS11 signing rig to test this patch, and I expect any missing deps to be caught as part of that.

**Comment by roliver-rpi** [comment]
> > Do we have any additional dependencies, e.g. `libengine-pkcs11-openssl`?
> 
> Unknown at this time - but that would make sense. I need to set up a PKCS11 signing rig to test this patch, and I expect any missing deps to be caught as part of that.

ACK

It looks like the SoftHSMv2 project (part of DNSSEC) implements the PKCS#11 interface entirely in software (backed by openssl). It should be possible to use openssl/libp11 to mux between various PKCS#11 implementations (including YubiKey's/SoftHSMv2). If so, we could potentially test these new code paths without requiring a signing rig.

**Comment by tdewey-rpi** [comment]
Merging under the testing I've been able to do - I want to get community feedback on this feature, as SoftHSM2 has been an absolute pain to get set up.

**Pull request merged**